File: backend/ProjectProgressManagementSystem/Services/Implementations/UserService.cs
using Microsoft.EntityFrameworkCore;
using ProjectProgressManagementSystem.DataAccess;
using ProjectProgressManagementSystem.Exceptions;
using ProjectProgressManagementSystem.Models;
using ProjectProgressManagementSystem.Services.Interfaces;
using ProjectProgressManagementSystem.Maps.ModelMappers;
using ProjectProgressManagementSystem.ViewModels.User;
using ProjectProgressManagementSystem.ViewModels;
using ProjectProgressManagementSystem.Utilities;
using System.Data;

namespace ProjectProgressManagementSystem.Services.Implementations
{
    public class UserService : IUserService
    {
        private readonly ApplicationDbContext _context;
        public UserService(ApplicationDbContext context)
        {
            _context = context;
        }

        // get all users
        public async Task<IEnumerable<UserViewModel>> GetAllAsync(Department? department, Region? region)
        {
            return await _context.Users
                .Where(user => department == null || (department.Value & user.Department) > 0)
                .Where(user => region == null || (region.Value & user.Region) > 0)
                .Where(user => user.DateDeleted == null || user.DateDeleted >= DateTime.Now)
                .Select(user => new UserViewModel
                {
                    Id = user.Id,
                    EmpId = user.EmpId,
                    UserName = user.UserName,
                    FirstName = user.FirstName,
                    LastName = user.LastName,
                    Email = user.Email,
                    IsExternal = user.IsExternal,
                    IsPasswordProtected = user.PasswordHash != null,
                    Department = user.Department,
                    Region = user.Region,
                    Role = user.Role,
                    WorkHoursPerDay = user.WorkHoursPerDay,
                    ParentId = user.ParentId,
                }).ToListAsync();
        }

        // get all users
        public async Task<IEnumerable<UserViewModel>> GetManagersAsync(Department? department, Region? region)
        {
            return await _context.Users
                .Where(user => department == null || (department.Value & user.Department) > 0)
                .Where(user => region == null || (region.Value & user.Region) > 0)
                .Where(user => user.DateDeleted == null || user.DateDeleted >= DateTime.Now)
                .Where(user => user.Role != Role.Employee && user.Role != Role.Developer)
                .Select(user => new UserViewModel
                {
                    Id = user.Id,
                    EmpId = user.EmpId,
                    UserName = user.UserName,
                    FirstName = user.FirstName,
                    LastName = user.LastName,
                    Email = user.Email,
                    IsExternal = user.IsExternal,
                    IsPasswordProtected = user.PasswordHash != null,
                    Department = user.Department,
                    Region = user.Region,
                    Role = user.Role,
                    WorkHoursPerDay = user.WorkHoursPerDay,
                    ParentId = user.ParentId,
                }).ToListAsync();
        }

        // get all users in month, year
        public async Task<IEnumerable<UserViewModel>> GetAllAsync(int year, int month, Department? department, Region? region)
        {
            var firstDayOfNextMonth = new DateTime(year, month, 1).AddMonths(1);
            return await _context.Users
                .Where(user => department == null || (department.Value & user.Department) > 0)
                .Where(user => region == null || (region.Value & user.Region) > 0)
                .Where(user => user.DateDeleted == null || user.DateDeleted >= firstDayOfNextMonth)
                .Where(user => user.DateCreated < firstDayOfNextMonth)
                .Select(user => new UserViewModel
                {
                    Id = user.Id,
                    EmpId = user.EmpId,
                    UserName = user.UserName,
                    FirstName = user.FirstName,
                    LastName = user.LastName,
                    Email = user.Email,
                    IsExternal = user.IsExternal,
                    IsPasswordProtected = user.PasswordHash != null,
                    Department = user.Department,
                    Region = user.Region,
                    Role = user.Role,
                    WorkHoursPerDay = user.WorkHoursPerDay,
                    ParentId = user.ParentId,
                }).ToListAsync();
        }

        // get all users under the same parent in month, year
        public async Task<IEnumerable<UserViewModel>> GetAllAsync(int year, int month, int parentId, Department? department, Region? region)
        {
            var firstDayOfNextMonth = new DateTime(year, month, 1).AddMonths(1);
            return await _context.Users
                .Where(user => department == null || (department.Value & user.Department) > 0)
                .Where(user => region == null || (region.Value & user.Region) > 0)
                .Where(user => user.DateDeleted == null || user.DateDeleted >= firstDayOfNextMonth)
                .Where(user => user.DateCreated < firstDayOfNextMonth && user.ParentId == parentId)
                .Select(user => new UserViewModel
                {
                    Id = user.Id,
                    EmpId = user.EmpId,
                    UserName = user.UserName,
                    FirstName = user.FirstName,
                    LastName = user.LastName,
                    Email = user.Email,
                    IsExternal = user.IsExternal,
                    IsPasswordProtected = user.PasswordHash != null,
                    Department = user.Department,
                    Region = user.Region,
                    Role = user.Role,
                    WorkHoursPerDay = user.WorkHoursPerDay,
                    ParentId = user.ParentId,
                }).ToListAsync();
        }

        // get user by id
        public async Task<UserViewModel> GetByIdAsync(int id)
        {
            var user = await FromId(id);
            return ViewModelMapper.ToViewModel(user);
        }

        // create a new user
        public async Task<UserViewModel> CreateAsync(UserCreateViewModel user)
        {
            var existingUser = await _context.Users.AnyAsync(u => u.DateDeleted == null && ((user.EmpId != null && u.EmpId == user.EmpId) || u.Email == user.Email));
            if (existingUser)
            {
                throw new DuplicateEntityException("User already exists");
            }

            var parent = await _context.Users.AnyAsync(u => u.Id == user.ParentId);
            if (user.ParentId != 0 && !parent)
            {
                throw new OperationNotSupportedException("Parent does not exists");
            }

            var userEntity = EntityModelMapper.ToEntity(user);
            await _context.Users.AddAsync(userEntity);
            await _context.SaveChangesAsync();
            return ViewModelMapper.ToViewModel(userEntity);
        }

        // import users from excel
        public async Task<MessageViewModel> ImportFromExcelAsync(IFormFile excelFile)
        {
            try
            {
                // Validate the file
                if (excelFile == null || excelFile.Length == 0)
                {
                    throw new ArgumentException("Excel file is required.");
                }

                var dataTable = ExcelReader.ReadExcelFile(excelFile);
                var users = new List<User>();

                foreach (DataRow row in dataTable.Rows)
                {
                    var user = EntityModelMapper.ToUserEntity(row);
                    users.Add(user);
                }
                // Add users to the database
                await _context.Users.AddRangeAsync(users);
                await _context.SaveChangesAsync();

                return new MessageViewModel
                {
                    Message = "Users imported successfully."
                };
            }
            catch (Exception ex)
            {
                throw new Exception($"Error importing users from Excel: {ex.Message}");
            }
        }

        // update an existing user
        public async Task<UserViewModel> UpdateAsync(int id, UserUpdateViewModel user)
        {
            var userEntity = await FromId(id);
            if (userEntity.DateDeleted != null)
            {
                throw new OperationNotSupportedException(
                    $"User with id {id} is scheduled to delete on {userEntity.DateDeleted}.");
            }
            userEntity.EmpId = user.EmpId ?? userEntity.EmpId;
            userEntity.UserName = user.UserName ?? userEntity.UserName;
            userEntity.FirstName = user.FirstName ?? userEntity.FirstName;
            userEntity.LastName = user.LastName ?? userEntity.LastName;
            userEntity.Email = user.Email ?? userEntity.Email;
            userEntity.IsExternal = user.Email?.Contains("ext") ?? userEntity.IsExternal;
            userEntity.Department = user.Department ?? userEntity.Department;
            userEntity.Region = user.Region ?? userEntity.Region;
            userEntity.Role = user.Role ?? userEntity.Role;
            userEntity.WorkHoursPerDay = user.WorkHoursPerDay ?? userEntity.WorkHoursPerDay;
            userEntity.ParentId = user.ParentId ?? userEntity.ParentId;
            userEntity.DateModified = DateTime.Now.Date;
            await _context.SaveChangesAsync();
            return ViewModelMapper.ToViewModel(userEntity);
        }

        // update last Saved Time for an existing user
        public async Task<UserViewModel> UpdateLastSavedTimeAsync(int id, DateTime lastSavedTime)
        {
            var userEntity = await FromId(id);
            userEntity.LastSavedTime = lastSavedTime;
            await _context.SaveChangesAsync();
            return ViewModelMapper.ToViewModel(userEntity);
        }

        // change password of an existing user
        public async Task<MessageViewModel> UpdatePasswordAsync(int id, PasswordCreateViewModel password)
        {
            var userEntity = await FromId(id);
            var passwordViewModel = EntityModelMapper.ToEntity(userEntity, password.OldPassword, password.NewPassword);
            userEntity.PasswordHash = passwordViewModel.PasswordHash;
            userEntity.PasswordSalt = passwordViewModel.PasswordSalt;
            await _context.SaveChangesAsync();
            return new MessageViewModel
            {
                Message = "Password changed successfully."
            };
        }

        // remove password of an existing user
        public async Task<MessageViewModel> RemovePasswordAsync(int id, string password)
        {
            var userEntity = await FromId(id);
            if (!EntityModelMapper.VerifyPasswordHash(password, userEntity.PasswordHash, userEntity.PasswordSalt))
            {
                throw new PasswordMismatchException();
            }
            userEntity.PasswordHash = null;
            userEntity.PasswordSalt = null;
            await _context.SaveChangesAsync();
            return new MessageViewModel
            {
                Message = "Password removed successfully."
            };
        }

        // delete a user
        public async Task<MessageViewModel> DeleteAsync(int id, bool? deleteNow)
        {
            var user = await FromId(id);

            // If deleteNow is not true, schedule the user for deletion on the first day of the next month.
            if (deleteNow != true)
            {
                // If the user is already scheduled for deletion, throw an exception.
                if (user.DateDeleted.HasValue)
                {
                    throw new OperationNotSupportedException($"User with id {id} is already scheduled to delete on {user.DateDeleted}.");
                }

                // Set the DateDeleted field to the first day of the next month.
                user.DateDeleted = new DateTime(DateTime.Now.Year, DateTime.Now.Month, 1).AddMonths(1);
            }
            else if (user.DateDeleted.HasValue)
            {
                // If deleteNow is true and the user is scheduled for deletion, cancel the deletion.
                user.DateDeleted = null;
            }

            // Remove the user from the Users DbSet. This will mark the user as deleted in the ChangeTracker.
            _context.Users.Remove(user);
            await _context.SaveChangesAsync();
            return new MessageViewModel
            {
                Message = $"User with id {id} is deleted successfully."
            };
        }

        // reset users table
        public async Task<MessageViewModel> ResetAsync()
        {
            await _context.Users.ExecuteDeleteAsync();
            return new MessageViewModel
            {
                Message = "Users table reset successfully."
            };
        }

        private async Task<User> FromId(int id)
        {
            return await _context.Users.FirstOrDefaultAsync(user => (user.DateDeleted == null || user.DateDeleted >= DateTime.Now) && user.Id == id)
                ?? throw new RecordNotFoundException($"Could not find any User with id: {id}");
        }
    }
}

File: backend/app/services/user_service.py
from datetime import date, datetime, timezone

from sqlalchemy import delete, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.exceptions import (
    DuplicateEntityException,
    OperationNotSupportedException,
    PasswordMismatchException,
    RecordNotFoundException,
)
from app.core.security import create_password_hash, verify_password
from app.models.enums import Department, Region, Role
from app.models.user import User
from app.schemas.common import MessageResponse
from app.schemas.user import PasswordChange, UserCreate, UserResponse, UserUpdate
from app.utils.mapper import user_to_entity, user_to_response


class UserService:
    def __init__(self, db: AsyncSession) -> None:
        self._db = db

    # ------------------------------------------------------------------ queries

    async def get_all(self, department: Department | None, region: Region | None) -> list[UserResponse]:
        stmt = (
            select(User)
            .where(
                (User.department.op("&")(int(department))) > 0 if department is not None else True,
                (User.region.op("&")(int(region))) > 0 if region is not None else True,
                (User.date_deleted.is_(None)) | (User.date_deleted >= datetime.now(timezone.utc)),
            )
        )
        users = (await self._db.execute(stmt)).scalars().all()
        return [user_to_response(u) for u in users]

    async def get_managers(self, department: Department | None, region: Region | None) -> list[UserResponse]:
        stmt = (
            select(User)
            .where(
                (User.department.op("&")(int(department))) > 0 if department is not None else True,
                (User.region.op("&")(int(region))) > 0 if region is not None else True,
                (User.date_deleted.is_(None)) | (User.date_deleted >= datetime.now(timezone.utc)),
                User.role.notin_([Role.Employee, Role.Developer]),
            )
        )
        users = (await self._db.execute(stmt)).scalars().all()
        return [user_to_response(u) for u in users]

    async def get_all_for_period(
        self,
        year: int,
        month: int,
        department: Department | None,
        region: Region | None,
        parent_id: int | None = None,
    ) -> list[UserResponse]:
        from app.utils.date_utils import first_day_of_next_month

        next_month = first_day_of_next_month(year, month)
        stmt = select(User).where(
            (User.department.op("&")(int(department))) > 0 if department is not None else True,
            (User.region.op("&")(int(region))) > 0 if region is not None else True,
            (User.date_deleted.is_(None)) | (User.date_deleted >= next_month),
            User.date_created < next_month,
        )
        if parent_id is not None:
            stmt = stmt.where(User.parent_id == parent_id)
        users = (await self._db.execute(stmt)).scalars().all()
        return [user_to_response(u) for u in users]

    async def get_by_id(self, user_id: int) -> UserResponse:
        user = await self._from_id(user_id)
        return user_to_response(user)

    async def get_by_email(self, email: str) -> UserResponse:
        stmt = select(User).where(User.email == email)
        user = (await self._db.execute(stmt)).scalar_one_or_none()
        if user is None:
            raise RecordNotFoundException(f"Could not find any User with email: {email}")
        return user_to_response(user)

    # ------------------------------------------------------------------ writes

    async def create(self, data: UserCreate) -> UserResponse:
        stmt = select(User).where(
            User.date_deleted.is_(None),
            (User.emp_id == data.emp_id) | (User.email == data.email),
        )
        existing = (await self._db.execute(stmt)).scalar_one_or_none()
        if existing:
            raise DuplicateEntityException("User already exists")

        if data.parent_id != 0:
            parent_stmt = select(User).where(User.id == data.parent_id)
            parent = (await self._db.execute(parent_stmt)).scalar_one_or_none()
            if parent is None:
                raise OperationNotSupportedException("Parent does not exist")

        user = user_to_entity(data)
        self._db.add(user)
        await self._db.flush()
        await self._db.refresh(user)
        return user_to_response(user)

    async def update(self, user_id: int, data: UserUpdate) -> UserResponse:
        user = await self._from_id(user_id)
        if user.date_deleted is not None:
            raise OperationNotSupportedException(
                f"User with id {user_id} is scheduled to delete on {user.date_deleted}."
            )
        if data.emp_id is not None:
            user.emp_id = data.emp_id
        if data.user_name is not None:
            user.user_name = data.user_name
        if data.first_name is not None:
            user.first_name = data.first_name.title()
        if data.last_name is not None:
            user.last_name = data.last_name.title()
        if data.email is not None:
            user.email = data.email
            user.is_external = "ext" in data.email
        if data.department is not None:
            user.department = int(data.department)
        if data.region is not None:
            user.region = int(data.region)
        if data.role is not None:
            user.role = int(data.role)
        if data.work_hours_per_day is not None:
            user.work_hours_per_day = data.work_hours_per_day
        if data.parent_id is not None:
            user.parent_id = data.parent_id
        if data.last_saved_time is not None:
            user.last_saved_time = data.last_saved_time
        user.date_modified = date.today()
        await self._db.flush()
        await self._db.refresh(user)
        return user_to_response(user)

    async def update_last_saved_time(self, user_id: int, last_saved_time: datetime) -> UserResponse:
        user = await self._from_id(user_id)
        user.last_saved_time = last_saved_time
        await self._db.flush()
        await self._db.refresh(user)
        return user_to_response(user)

    async def change_password(self, user_id: int, data: PasswordChange) -> MessageResponse:
        user = await self._from_id(user_id)
        if data.old_password is not None and not verify_password(
            data.old_password, user.password_hash, user.password_salt
        ):
            raise PasswordMismatchException()
        if data.old_password == data.new_password:
            raise ValueError("Old and New Password are the same.")
        password_hash, password_salt = create_password_hash(data.new_password)
        user.password_hash = password_hash
        user.password_salt = password_salt
        await self._db.flush()
        return MessageResponse(message="Password changed successfully.")

    async def remove_password(self, user_id: int, current_password: str) -> MessageResponse:
        user = await self._from_id(user_id)
        if not verify_password(current_password, user.password_hash, user.password_salt):
            raise PasswordMismatchException()
        user.password_hash = None
        user.password_salt = None
        await self._db.flush()
        return MessageResponse(message="Password removed successfully.")

    async def delete(self, user_id: int, delete_now: bool | None) -> MessageResponse:
        from app.utils.date_utils import first_day_of_next_month

        user = await self._from_id(user_id)
        if delete_now is not True:
            if user.date_deleted is not None:
                raise OperationNotSupportedException(
                    f"User with id {user_id} is already scheduled to delete on {user.date_deleted}."
                )
            today = date.today()
            user.date_deleted = first_day_of_next_month(today.year, today.month)
        elif user.date_deleted is not None:
            user.date_deleted = None
        await self._db.delete(user)
        await self._db.flush()
        return MessageResponse(message=f"User with id {user_id} is deleted successfully.")

    async def reset(self) -> MessageResponse:
        await self._db.execute(delete(User))
        await self._db.flush()
        return MessageResponse(message="Users table reset successfully.")

    # ------------------------------------------------------------------ helpers

    async def _from_id(self, user_id: int) -> User:
        stmt = select(User).where(
            (User.date_deleted.is_(None)) | (User.date_deleted >= datetime.now(timezone.utc)),
            User.id == user_id,
        )
        user = (await self._db.execute(stmt)).scalar_one_or_none()
        if user is None:
            raise RecordNotFoundException(f"Could not find any User with id: {user_id}")
        return user
